### DISCUSSION TITLE
Problem with Backed enums

### DISCUSSION THREAD

**Discussion by rabol** [larastan/larastan#2096] [Q&A]
Hi

I'm struggeling to make Larastan happy with enums.

```
$userDoc->status->getLabel()
```

gives an error like this:

```
 Cannot call method getLabel() on int.
```

enum looks like this:
```
<?php

declare(strict_types=1);

namespace App\Enums;

use App\Traits\EnumHelper;
use Filament\Support\Contracts\HasLabel;

enum UserDocStatusEnum: int implements HasLabel
{
    case None = 0;
    case Created = 10;
    case Draft = 20;
    case SentToReceiver = 30;
    case SenderHasSigned = 31;
    case ReceiverOrRecipientHasSigned = 32;

    case Rejected = 40;
    case SendToRecipients = 50;
    case SenderToSign = 60;
    case PendingDownload = 70;

    case SenderDownloaded = 79;

    case Completed = 80;

    public function getLabel(): ?string
    {
        return match ($this) {
            self::None => __('None'),
            self::Created => __('Created'),
            self::Draft => __('Draft'),
            self::SentToReceiver => __('Sent to receiver'),
            self::SenderHasSigned => __('Sender has signed'),
            self::ReceiverOrRecipientHasSigned => __('Receiver or Recipient has signed'),
            self::Rejected => __('Receiver rejected'),
            self::SendToRecipients => __('Sent to recipient'),
            self::SenderToSign => __('Sender to sign'),
            self::PendingDownload => __('Pending download'),
            self::SenderDownloaded => __('Sender has downloaded'),
            self::Completed => __('Completed'),
        };
    }
}
```

This example works without errors in Tinker/Tinkerwell

```
use App\Models\UserDoc;

$userDoc = UserDoc::first();
echo $userDoc->status->getLabel();
```

So how can I make Larastan happy ?

**Comment by aniruddhaadak80** [comment]
Hi @rabol ,

The error occurs because Larastan interprets `$userDoc->status` as an `int` rather than the `UserDocStatusEnum`. To resolve this:

1. **Explicit Type Casting:** Ensure the `status` attribute on the `UserDoc` model is explicitly cast as an `enum`. In Laravel, add the cast in the model:

    ```php
    protected $casts = [
        'status' => UserDocStatusEnum::class,
    ];
    ```

2. **Static Analysis Tip for Larastan:** Sometimes Larastan may still throw warnings even with casting. To address this, you can use `@var` annotations to clarify the variable type for Larastan, like so:

    ```php
    /** @var UserDocStatusEnum $status */
    echo $userDoc->status->getLabel();
    ```

This should prevent Larastan from misinterpreting `status` as an `int` and allow it to recognize the `getLabel()` method.

Thank you.

**Reply by rabol** [reply]
Hello,

Sorry I forgot to say that my model have this:

```
protected function casts(): array
    {
        return [
            'add_sender_signature' => 'bool',
            'add_sender_signature_last' => 'bool',
            'receiver_signed_at' => 'datetime',
            'sender_signed_at' => 'datetime',
            'status' => \App\Enums\UserDocStatusEnum::class,
            'sender_download' => 'integer',
            'send_only' => 'bool',
            'need_sender_signature' => 'bool',
            'send_download_link' => 'bool',
            'is_in_person' => 'bool',
            'custom_data' => 'array',
            'receiver_custom_data' => 'array',
            'send_doc_when_all_has_signed' => 'bool',
            'need_id_validation' => 'bool',
            'pdf_encryptMetadata' => 'bool',
            'pdf_permissions' => 'array',
        ];
    }
```

I have tried to add:
```
/** @var UserDocStatusEnum $status */
$this->Cell($w4, 5, $userDoc->status->getLabel(), 'B');
```

But I stil get the same error

**Reply by aniruddhaadak80** [reply]
Since you’ve confirmed that `status` is cast to `UserDocStatusEnum` in your model, the issue is likely due to Larastan’s strict type inference, which sometimes doesn’t recognize enum-backed types automatically.

### Solution Adjustments:

1. **Type Hint on Access**:
   Add a specific type hint where you access `getLabel()`:
   ```php
   /** @var \App\Enums\UserDocStatusEnum $userDoc->status */
   $this->Cell($w4, 5, $userDoc->status->getLabel(), 'B');
   ```
   This approach tells Larastan explicitly that `status` is a `UserDocStatusEnum`.

2. **Casting as an Enum**:
   Another approach is to cast `status` as `UserDocStatusEnum` directly when calling `getLabel()`:
   ```php
   $this->Cell($w4, 5, ($userDoc->status instanceof UserDocStatusEnum ? $userDoc->status : UserDocStatusEnum::from($userDoc->status))->getLabel(), 'B');
   ```

3. **Add `@phpstan-ignore`** (Temporary):
   If the issue persists, using `@phpstan-ignore-next-line` above the line will suppress the error as a last resort. 

These should help Larastan recognize the enum type and resolve the error.

**Reply by rabol** [reply]
Hi

I ended up with

```
   /** @phpstan-ignore-next-line */
```

But IMHO, that is not the solution.

**Reply by aniruddhaadak80** [reply]
Hi @rabol ,

Here are the two most important solutions:

1. **Explicit Type Hinting**: Ensure that you provide a type hint for the `$userDoc->status` variable. This helps PHPStan understand that it's an instance of `UserDocStatusEnum`. You can do this by explicitly defining it in your PHPDoc:
   ```php
   /** @var UserDocStatusEnum $status */
   $status = $userDoc->status;
   $this->Cell($w4, 5, $status->getLabel(), 'B');
   ```

2. **Enhance Model Casting**: Ensure that your model properly casts the `status` attribute to the `UserDocStatusEnum`. In your `casts()` method, confirm that the enum class is correctly specified, which helps Larastan recognize it as an enum type:
   ```php
   protected function casts(): array
   {
       return [
           'status' => \App\Enums\UserDocStatusEnum::class,
           // other casts...
       ];
   }
   ```

Thanks.

**Comment by szepeviktor** [comment]
Larastan may not fully support backed enums.

**Comment by calebdw** [comment]
Sorry folks, all the suggestions in this discussion are off---Larastan supports backed enums and there's no need to locally cast it to an enum or use `@var`/`@phpstan-ignore`.

The problem is that you're not defining your casts properly, you either need to:
1. Use the `$casts` property
2. Add the `@return` phpdoc to the `casts()` method:
    ```php
    /** 
     * @return array{
     *     status: 'App\\Enums\\UserDocStatusEnum',
     *     ...
     * }
     */
    protected function casts(): array
    {
        return [
            'status' => \App\Enums\UserDocStatusEnum::class,
            // other casts...
        ];
    }
    ```

See #1878 and https://github.com/larastan/larastan/blob/2.x/UPGRADE.md#upgrading-to-296-from-295 for more details.

**Reply by rabol** [reply]
Thanks for your input, but I have the casts and the return like this:

```
 /**
     * @return array{
     *  add_sender_signature: 'bool',
     *  add_sender_signature_last:  'bool',
     *  receiver_signed_at: 'datetime',
     *  sender_signed_at: 'datetime',
     *  status: 'App\Enums\UserDocStatusEnum',
     *  sender_download: 'integer',
     *  send_only: 'bool',
     *  need_sender_signature: 'bool',
     *  send_download_link: 'bool',
     *  is_in_person: 'bool',
     *  custom_data: 'array',
     *  receiver_custom_data: 'array',
     *  send_doc_when_all_has_signed: 'bool',
     *  need_id_validation:  'bool',
     *  pdf_encryptMetadata: 'bool',
     *  pdf_permissions: 'array'
     * }
     */
```

**Reply by calebdw** [reply]
Ah, your cast methods need to be `public`. Duplicate of #2087

**Reply by rabol** [reply]
does not make a difference

**Reply by calebdw** [reply]
You'll need to clear your cache and rerun:

```bash
phpstan clear-result-cache
```

**Reply by rabol** [reply]
Same result - do I really need to have everything in one line ?

**Reply by calebdw** [reply]
I think you need two slashes in your string:
```
status: 'App\\Enums\\UserDocStatusEnum'

```

**Reply by rabol** [reply]
```
Method App\Models\UserDoc::casts() should return array{add_sender_signature: 'bool', add_sender_signature_last: 'bool', receiver_signed_at: 'datetime', sender_signed_at: 'datetime', status: 'App\\\\Enums\\…', sender_download: 'integer', send_only: 'bool', need_sender_signature: 'bool', ...} but returns array{add_sender_signature: 'bool', add_sender_signature_last: 'bool', receiver_signed_at:  
         'datetime', sender_signed_at: 'datetime', status: 'App\\Enums\\UserDocStatusEnum', sender_download: 'integer', send_only: 'bool', need_sender_signature: 'bool', ...}.                                                                                                                                                                                                                                      
         💡 Offset 'status' ('App\\\\Enums\\…') does not accept type 'App\\Enums\\UserDocStatusEnum'.  ```

**Reply by rabol** [reply]
I might have found something:

one cannot have this: 
```
treatPhpDocTypesAsCertain: false
```

in phpstan.neon
